File: 03/null_pointer.rb
class Greeter
  def greet(n)
    return 'Hello' if n > 0       # <1>
  end
end

class Responder
  def answer_greeting(klass)
    greeting = klass.greet(0)
    if greeting.match(/[Hh]ello/) # <2>
      'Nice to meet you!'
    else
      'Um, Hello?'
    end
  end
end

class NullPointer
  def self.run
    greeter = Greeter.new
    s = Responder.new.answer_greeting(greeter)
    puts s
  end
end

NullPointer.run

File: 05/NullPointer.java
public class NullPointer {

  static class Greeter {
    String greet(int n) {                   // <1>
      if (n > 0) {
        return "Hello";                     // <2>
      } else {
        return null;                        // <3>
      }
    }
  }

  static class Responder {
    String answerGreeting(Greeter klass) {
      String greeting = klass.greet(0);     // <4>
      if (greeting.equals("Hello")) {
        return "Nice to meet you!";
      } else {
        return "Um, Hello?";
      }
    }
  }

  public static void main(String args[]) {
    Greeter greeter = new Greeter();
    Responder other = new Responder();

    String s = other.answerGreeting(greeter);  // <5>

    //int x = null;  // not okay, this is a compile error
    System.out.println(s);
  }
}
